File: src/chatbot_pipeline/pipeline.py
from __future__ import annotations

from pathlib import Path

from .config import Settings
from .documents import copy_to_uploads, prepare_for_pageindex
from .guardrails import build_messages, enforce_grounding
from .pageindex_gateway import PageIndexGateway
from .state import DocumentRecord, ManifestStore


class GroundedChatbotPipeline:
    def __init__(self, settings: Settings) -> None:
        self.settings = settings
        self.store = ManifestStore(settings.manifest_path)
        self.pageindex = PageIndexGateway(settings.pageindex_api_key)

    def ingest(self, input_path: Path, wait: bool = True) -> DocumentRecord:
        uploaded_path = copy_to_uploads(input_path, self.settings.upload_dir)
        submitted_path = prepare_for_pageindex(uploaded_path, self.settings.converted_dir)
        doc_id = self.pageindex.submit_document(submitted_path)
        record = DocumentRecord.new(
            doc_id=doc_id,
            original_name=input_path.name,
            uploaded_path=uploaded_path,
            submitted_path=submitted_path,
        )
        self.store.upsert(record)

        if wait:
            record.status = self.pageindex.wait_until_complete(
                doc_id,
                timeout_seconds=self.settings.processing_timeout_seconds,
                poll_interval_seconds=self.settings.poll_interval_seconds,
            )
            self.store.upsert(record)
        return record

    def refresh_status(self, doc_id: str) -> str:
        status = self.pageindex.get_document_status(doc_id)
        self.store.update_status(doc_id, status)
        return status

    def ask(self, question: str, doc_ids: list[str] | None = None) -> str:
        scoped_doc_ids = doc_ids or self.store.completed_ids()
        if not scoped_doc_ids:
            return "I don't know"
        raw_answer = self.pageindex.chat(build_messages(question), scoped_doc_ids)
        return enforce_grounding(raw_answer, self.settings.require_pageindex_citations)

    def documents(self) -> list[DocumentRecord]:
        return self.store.list()

File: src/pipeline.ts
import { copyToUploads, prepareForPageIndex } from "./documents.js";
import { buildMessages, enforceGrounding, IDK_ANSWER } from "./guardrails.js";
import { PageIndexGateway } from "./pageindexGateway.js";
import { ManifestStore, newDocumentRecord, type DocumentRecord } from "./state.js";
import type { Settings } from "./config.js";

export type IngestOptions = {
  wait?: boolean;
  originalName?: string;
};

export class GroundedChatbotPipeline {
  private readonly store: ManifestStore;
  private pageIndex?: PageIndexGateway;

  constructor(private readonly settings: Settings) {
    this.store = new ManifestStore(settings.manifestPath);
  }

  async ingest(inputPath: string, options: IngestOptions = {}): Promise<DocumentRecord> {
    const wait = options.wait ?? true;
    const originalName = options.originalName ?? inputPath.split(/[\\/]/).at(-1) ?? "document";
    const uploadedPath = await copyToUploads(inputPath, this.settings.uploadDir, originalName);
    const submittedPath = await prepareForPageIndex(uploadedPath, this.settings.convertedDir);
    const docId = await this.getPageIndex().submitDocument(submittedPath);

    const record = newDocumentRecord(docId, originalName, uploadedPath, submittedPath);
    await this.store.upsert(record);

    if (wait) {
      record.status = await this.getPageIndex().waitUntilComplete(
        docId,
        this.settings.processingTimeoutSeconds,
        this.settings.pollIntervalSeconds
      );
      await this.store.upsert(record);
    }

    return record;
  }

  async refreshStatus(docId: string): Promise<string> {
    const status = await this.getPageIndex().getDocumentStatus(docId);
    await this.store.updateStatus(docId, status);
    return status;
  }

  async ask(question: string, docIds?: string[]): Promise<string> {
    const scopedDocIds = docIds?.length ? docIds : await this.store.completedIds();
    if (!scopedDocIds.length) return IDK_ANSWER;

    const rawAnswer = await this.getPageIndex().chat(buildMessages(question), scopedDocIds);
    return enforceGrounding(rawAnswer, this.settings.requirePageIndexCitations);
  }

  async documents(): Promise<DocumentRecord[]> {
    return this.store.list();
  }

  private getPageIndex(): PageIndexGateway {
    this.pageIndex ??= new PageIndexGateway(this.settings.pageIndexApiKey);
    return this.pageIndex;
  }
}
